pkg/agent/clean/active_directory.go

```diff
@@ -0,0 +1,1100 @@
+			}
+		} else if index, exists = duplicateGUIDWorkUnits[prtb.UserPrincipalName]; exists {
+			if workUnitContainsName(&(*workunits)[index], prtb.UserName) {
+				(*workunits)[index].duplicateLocalPRTBs = append((*workunits)[index].duplicateLocalPRTBs, prtb)
+			} else {
+				logrus.Warnf("[%v] found PRTB for user with guid-based principal '%v' and name '%v', but no user object with that name matches the GUID or its associated DN. refusing to process",
+					identifyAdUserOperation, prtb.UserPrincipalName, prtb.UserName)
+			}
+		}
+	}
+
+	return nil
+}
+
+func migrateCRTBs(workunit *migrateUserWorkUnit, sc *config.ScaledContext, dryRun bool) error {
+	crtbInterface := sc.Management.ClusterRoleTemplateBindings("")
+	// First convert all GUID-based CRTBs to their equivalent Distinguished Name variants
+	dnPrincipalID := activeDirectoryPrefix + workunit.distinguishedName
+	for _, oldCrtb := range workunit.guidCRTBs {
+		if dryRun {
+			logrus.Infof("[%v] DRY RUN: would migrate CRTB '%v' from GUID principal '%v' to DN principal '%v'. "+
+				"Additionally, an annotation, %v, would be added containing the principal being migrated from and"+
+				"labels, %v and %v, that will contain the name of the previous CRTB and indicate that this CRTB has been migrated.",
+				migrateCrtbsOperation, oldCrtb.Name, oldCrtb.UserPrincipalName, dnPrincipalID, adGUIDMigrationAnnotation, migrationPreviousName, adGUIDMigrationLabel)
+		} else {
+			newAnnotations := oldCrtb.Annotations
+			if newAnnotations == nil {
+				newAnnotations = make(map[string]string)
+			}
+			newAnnotations[adGUIDMigrationAnnotation] = oldCrtb.UserPrincipalName
+			newLabels := oldCrtb.Labels
+			if newLabels == nil {
+				newLabels = make(map[string]string)
+			}
+			newLabels[migrationPreviousName] = oldCrtb.Name
+			newLabels[adGUIDMigrationLabel] = migratedLabelValue
+			newCrtb := &v3.ClusterRoleTemplateBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:         "",
+					Namespace:    oldCrtb.ObjectMeta.Namespace,
+					GenerateName: "crtb-",
+					Annotations:  newAnnotations,
+					Labels:       newLabels,
+				},
+				ClusterName:       oldCrtb.ClusterName,
+				UserName:          workunit.originalUser.Name,
+				RoleTemplateName:  oldCrtb.RoleTemplateName,
+				UserPrincipalName: dnPrincipalID,
+			}
+			_, err := crtbInterface.Create(newCrtb)
+			if err != nil {
+				return fmt.Errorf("[%v] unable to create new CRTB: %w", migrateCrtbsOperation, err)
+			}
+			err = sc.Management.ClusterRoleTemplateBindings("").DeleteNamespaced(oldCrtb.Namespace, oldCrtb.Name, &metav1.DeleteOptions{})
+			if err != nil {
+				return fmt.Errorf("[%v] unable to delete CRTB: %w", migrateCrtbsOperation, err)
+			}
+		}
+	}
+	// Now do the same for Local ID bindings on the users we are about to delete, pointing them instead to the merged
+	// original user that we will be keeping
+	localPrincipalID := localPrefix + workunit.originalUser.Name
+	for _, oldCrtb := range workunit.duplicateLocalCRTBs {
```

Comment:
Here (and for the other types) the logic for duplicate and guid resources is very similar. I would recommend moving this into a side function, and passing slight different parameters to it as needed. Should reduce code complexity.